pkg/randchar/generator.go

```diff
@@ -1,54 +1,306 @@
+// which panics upon error so be careful. For more safety, use NewRand instead.
+func MustNewRand(base Charset, options ...Option) Rand {
+	r, err := NewRand(base, options...)
+	if err != nil {
+		panic(err)
+	}
+	return r
+}
+
+// Option defines a configuration option for a random reader.
+type Option func(r Rand) (Rand, error)
+
+// minimum defines a minimum number of characters that must be from a given character set.
+type minimum struct {
+	count   int
+	charset Charset
+}
+
+// Min ensures the generated slice contains at least n characters from the given character set.
+// When multiple Min options are given with the same character set, the biggest minimum takes precedence.
+func Min(n int, charset Charset) Option {
+	return func(r Rand) (Rand, error) {
+		if n < 1 {
+			return r, errors.New("minimum must be at least 1")
+		}
+
+		if len(charset) == 0 {
+			return r, errors.New("minimum character set cannot be empty")
+		}
+
+		if !charset.IsSubset((r.base)) {
+			return r, errors.New("minimum character set must be a subset of the base character set")
+		}
+
+		// Ensure the biggest minimum takes precedence when same charset minimum applies.
+		for i, minimum := range r.minima {
+			if minimum.charset.Equal(charset) {
+				if minimum.count >= n {
+					return r, nil
+				}
+
+				r.minLen += n - minimum.count
+				r.minima[i].count = n
+				return r, nil
+			}
+		}
+
+		r.minima = append(r.minima, minimum{
+			count:   n,
+			charset: charset,
+		})
+		r.minLen += n
+		return r, nil
+	}
+}
+
+// WithReader allows you to set the reader used as source of randomness.
+// Do not use this unless you know what you're doing. By default, the
+// source of randomness is set to crypto/rand.Reader.
+func WithReader(reader io.Reader) Option {
+	return func(r Rand) (Rand, error) {
+		r.reader = reader
+		return r, nil
+	}
+}
+
+// Generate returns a randomly generated slice of characters that meets the requirements of the reader.
+func (r Rand) Generate(n int) ([]byte, error) {
+	if n < r.minLen {
+		return nil, errors.New("n cannot be smaller than the minimum required length of the generator")
+	}
+
+	if n < 1 {
+		return []byte{}, nil
+	}
+
+	result := make([]byte, n)
+	i := 0
+	var err error
+	for _, min := range r.minima {
+		err = min.charset.rand(r.reader, result[i:i+min.count])
+		if err != nil {
+			return nil, err
+		}
+		i += min.count
+	}
+
+	err = r.base.rand(r.reader, result[i:])
+	if err != nil {
+		return nil, err
+	}
+
+	// Only shuffle if we've applied the minima.
+	// Otherwise, the result is already random and
+	// we can save time by skipping the shuffle step.
+	if i > 0 {
+		err = shuffle(r.reader, result)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+// shuffle randomly shuffles elements of a byte slice, using the Durstenfeld shuffle algorithm.
+func shuffle(reader io.Reader, data []byte) error {
+	var randomIndex *big.Int
+	var j int64
+	var err error
+	for i := len(data) - 1; i > 0; i-- {
```

Comment:
Wow that could be dangerous! After some investigation, I think it **does** get swapped. 

For each element `i`, it is swapped with another randomly chosen element `j` where `0 <= j <= i`. This means that on each iteration through the `for` loop, `data[0]` has a chance to be selected as the element to swap with `i`. 

Good call to check it, though I think we're safe 😄 

For some background, check out these resources:

- Algorithm description: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm
- Also, the `math/rand` package seems to have implemented the Fisher-Yates algorithm (which is similar) for their `Shuffle` function: https://golang.org/src/math/rand/rand.go?s=11285:11325#L232 It's a bit less readable though.

Reply:
Aha, but I did identify a different bug! Turns out `rand.Int(reader, i)` produces a value `[0, i)`, which is _exclusive_ `i` and we need it to _inclusive_ `i`.

See eb2e164 for the fix.

Reply:
Ahh, I was thinking because `i` never reached 0, data[0] would never get swapped. However, as you point out `j` can reach 0. So data[0] can get swapped. :+1: 

`i > 0` is the correct boundary, as having `i=0` would make `j=0` (random integer in the [0,0] range). This wouldn't change anything, as you'd swap `data[0]` with `data[0]`.